### ISSUE TITLE
Jest unit tests are not respecting Next.js config when rendering

### ISSUE THREAD

**Issue by terrymun** [nrwl/nx#11948]
## Current Behavior

TL;DR: @nrwl/jest does not seem to be taking into account Next.js config when rendering for unit tests.

The [experimental flag `newNextLinkBehavior`](https://github.com/vercel/next.js/pull/36436) is not being used by Nx when rendering `<Link>` components in unit tests. It, however, renders the `<Link>` components correctly when serving or building the site.

This bug was discovered when attempting to future proof our local Nx + Next.js setup for next@13 that will be released in the near future. This experimental flag toggles the experimental rendering behavior of `<Link>`, where it will render an `<a>` in place without requiring a nested `<a>` child.

However, this flag does not seem to be respected when running Jest unit tests:

```tsx
<Link href="https://example.com"><strong>A link</strong></Link>
```

...will be rendered as:

```tsx
<strong>A link</strong>
```

This is apparent when checking the rendered output against inline snapshots:

```tsx
it("should render <Link> element with nested child with new behavior", () => {
  const { container } = render(
    <Link href="https://example.com" data-testid="link">
      <strong data-testid="content">Link element with new behavior</strong>
    </Link>
  );
  expect(container).toMatchInlineSnapshot(`
    <div>
      <strong
        data-testid="content"
      >
        Link element with new behavior
      </strong>
    </div>
  `);
  // Notice the incorrect markup rendered. The `<strong>` tag should have been wrapped in an `a` tag
  // If this TSX template is used on `index.tsx` then we will see correct output
});
```

## Expected Behavior

```
<Link href="https://example.com"><strong>A link</strong></Link>
```

...should be rendered as:

```
<a href="https://example.com"><strong>A link</strong></a>
```

The inline snapshot is expected to have an `<a>` wrapping the `<strong>` element, i.e.:

```tsx
it("should render <Link> element with nested child with new behavior", () => {
  const { container } = render(
    <Link href="https://example.com" data-testid="link">
      <strong data-testid="content">Link element with new behavior</strong>
    </Link>
  );
  expect(container).toMatchInlineSnapshot(`
    <div>
      <a href="https://example.com">
       <strong
          data-testid="content"
        >
          Link element with new behavior
        </strong>
      </a>
    </div>
  `);
});
```

This is verified to work when serving the app using `yarn next`. So the discrepancy here is that Jest is not rendering the markup correctly, but serving the site does result in correct rendering. This causes unit tests to fail, but end-to-end tests to pass.

*Question: Is this a regression? .i.e Did this used to be the behavior at one point?*

Not that I know of, as this flag is recently introduced, and it simply seems that Nx is not interpolating this setting from `next.config.ts` when serving or testing the app.

## Steps to Reproduce

A minimal Github repo is available at this link: https://github.com/terrymun/nx-nextjs-link-behavior-demo

1. Run `index.spec.tsx` and note the generated inline snapshot
2. Run `yarn start` and compare the generated markup: they are different from the inline snapshot, suggesting that Jest's render function has failed to take into account the experimental flag from `next.config.ts`3. 

### Failure Logs

n.a.

### Environment

```
Node : 16.16.0
OS   : darwin x64
npm  : 8.11.0

nx : 14.6.5
@nrwl/angular : Not Found
@nrwl/cypress : 14.6.5
@nrwl/detox : Not Found
@nrwl/devkit : 14.6.5
@nrwl/eslint-plugin-nx : 14.6.5
@nrwl/express : Not Found
@nrwl/jest : 14.6.5
@nrwl/js : 14.6.5
@nrwl/linter : 14.6.5
@nrwl/nest : Not Found
@nrwl/next : 14.6.5
@nrwl/node : Not Found
@nrwl/nx-cloud : 14.6.2
@nrwl/nx-plugin : Not Found
@nrwl/react : 14.6.5
@nrwl/react-native : Not Found
@nrwl/schematics : Not Found
@nrwl/storybook : 14.6.5
@nrwl/web : 14.6.5
@nrwl/workspace : 14.6.5
typescript : 4.7.4
---------------------------------------
Local workspace plugins:
---------------------------------------
Community plugins:
```

**Comment by barbados-clemens** [comment]
does this work with a plain next app + jest + @testing-library/react?
 
Initially I would assume this is just because @testing-library/react serialization isn't considering that change in the config.

Jest itself isn't doing the rendering here from my understanding. it's just serializing what it's passed.

**Comment by terrymun** [comment]
@barbados-clemens Thanks for the reply! It seems that using `next/jest` should be the way going forward (even though it seems that even using this recommended way still results in mismatched renders on a plain next app: [repro link](https://github.com/terrymun/nextjs-link-behavior-demo)).

However when trying to test this on Nx + Next.js, [using `next/js` as recommended](https://nextjs.org/docs/testing#setting-up-jest-with-the-rust-compiler) with the following update to `jest.config.ts`:

```
/* eslint-disable */
const nextJest = require('next/jest');

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
});

// Add any custom config to be passed to Jest
/** @type {import('jest').Config} */
const customJestConfig = {
  displayName: 'site',
  preset: '../../jest.preset.js',
  transform: {
    '^(?!.*\\.(js|jsx|ts|tsx|css|json)$)': '@nrwl/react/plugins/jest',
    '^.+\\.[tj]sx?$': ['babel-jest', { presets: ['@nrwl/next/babel'] }],
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/apps/site',
}

export default createJestConfig(customJestConfig);
```

...causes an error:

```
Error: Jest: Failed to parse the TypeScript config file /Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/apps/site/jest.config.ts
  Error: > Couldn't find a `pages` directory. Please create one under the project root
    at readConfigFileAndSetRootDir (/Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/node_modules/@jest/core/node_modules/jest-config/build/readConfigFileAndSetRootDir.js:136:13)
    at async readConfig (/Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/node_modules/@jest/core/node_modules/jest-config/build/index.js:208:18)
    at async readConfigs (/Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/node_modules/@jest/core/node_modules/jest-config/build/index.js:404:26)
    at async runCLI (/Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/node_modules/@jest/core/build/cli/index.js:140:59)
    at async Object.run (/Users/[REDACTED]/Documents/GitHub/nx-nextjs-link-behavior-demo/node_modules/jest-cli/build/cli/index.js:155:37)
```

This is likely a separate issue, so please let me know if I should create a new ticket/issue for this :) thanks again for taking your time to look into this!

**Comment by barbados-clemens** [comment]
hmm might be nextJest not understanding the monorepo setup or might need more information to understand how the app is configured or might get confused with the way nx jest is set up by default. I can take a look a little later this week to see if I come up with anything.

**Comment by barbados-clemens** [comment]
looking at the reproducible shows it working with the default setup except the nested one. (I updated the snapshots to match what the DOM should be displaying so the tests will fail accordingly)
is this what you're seeing @terrymun?
```
> nx run site:test

 FAIL   site  apps/site/specs/index.spec.tsx
  Index
    ✓ should render <Link> element with new behavior (12 ms)
    ✕ should render <Link> element with nested child with new behavior (4 ms)
    ✓ should render <Link> element with legacy behavior (3 ms)

  ● Index › should render <Link> element with nested child with new behavior

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `Index should render <Link> element with nested child with new behavior 1`

    - Snapshot  - 2
    + Received  + 0

      <div>
    - <a data-testid="link" href="https://example.com">
        <strong
          data-testid="content"
        >
          Link element with new behavior
        </strong>
    - </a>
      </div>

      28 |       </Link>
      29 |     );
    > 30 |     expect(container).toMatchInlineSnapshot(`
         |                       ^
      31 |       <div>
      32 |       <a data-testid="link" href="https://example.com">
      33 |         <strong

      at Object.<anonymous> (specs/index.spec.tsx:30:23)

 › 1 snapshot failed.
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or re-run jest with `-u` to update them.

Test Suites: 1 failed, 1 total
Tests:       1 failed, 2 passed, 3 total
Snapshots:   1 failed, 2 passed, 3 total
Time:        0.591 s, estimated 1 s
Ran all test suites.

 ——————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target test for project site (1s)

    ✖    1/1 failed
    ✔    0/1 succeeded [0 read from cache]

   See Nx Cloud run details at https://nx.app/runs/nWnGnQM0yD
```

**Comment by terrymun** [comment]
That is correct: the snapshot failing has to do with a separate issue with `next/jest` not working as expected.

Can you elaborate on the statement "working with the default setup except the nested one"? How did you get it to work?

**Comment by terrymun** [comment]
Many thanks @barbados-clemens! Again, really appreciate the time you took into looking. I think the gist of this PR is mostly me needing instructions on what to provide to the `dir` argument for `nextJest`.

The jest config issue is now also being fixed in nextjs canary, and I can confirm that the Link rendering bug is not a Nx issue at all.

Feel free to close this question, and thanks again for your time :)

**Comment by barbados-clemens** [comment]
not too sure, you could just make an empty pages directory and see if that gets past the error since you're not testing any pages. though with a lib you also don't have a next config (by default) in there so that could also be an issue? wonder if you can just say the directory is in the root of the app and then still used the libs jest config? 🤔

**Comment by terrymun** [comment]
@barbados-clemens Silly me, having an empty `/pages` directory worked.

Ideally yes, a libs folder should not have a next config in there, but if the lib is indeed re-exporting some next components (like `<Link>`) then it does need a next config so it will honour whatever experimental flag that is enabled.
